rust-package/brainflow/src/board_shim.rs

```diff
@@ -0,0 +1,519 @@
+use once_cell::sync::Lazy;
+use paste::paste;
+use std::{
+    ffi::CString,
+    os::raw::{c_double, c_int},
+    path::Path,
+    sync::Mutex,
+};
+
+use crate::{
+    brainflow_input_params::BrainFlowInputParams,
+    check_brainflow_exit_code,
+    error::{BrainFlowError, Error},
+    BoardId, Result,
+};
+
+use crate::ffi::board_controller::BoardController;
+use crate::ffi::constants::LogLevels;
+
+const MAX_CHANNELS: usize = 512;
+
+#[cfg(target_os = "windows")]
+pub static BOARD_CONTROLLER: Lazy<Mutex<BoardController>> = Lazy::new(|| {
+    #[cfg(target_pointer_width = "64")]
+    let lib_path = Path::new("lib\\libBoardController.dll");
+    #[cfg(target_pointer_width = "32")]
+    let lib_path = Path::new("lib\\libBoardController32.dll");
+    let board_controller = unsafe { BoardController::new(lib_path).unwrap() };
+    Mutex::new(board_controller)
+});
+
+#[cfg(target_os = "macos")]
+pub static BOARD_CONTROLLER: Lazy<Mutex<BoardController>> = Lazy::new(|| {
+    let lib_path = Path::new("lib/libBoardController.dylib");
+    let board_controller = unsafe { BoardController::new(lib_path).unwrap() };
+    Mutex::new(board_controller)
+});
+
+#[cfg(target_os = "linux")]
+pub static BOARD_CONTROLLER: Lazy<Mutex<BoardController>> = Lazy::new(|| {
+    let lib_path = Path::new("lib/libBoardController.so");
+    let board_controller = unsafe { BoardController::new(lib_path).unwrap() };
+    Mutex::new(board_controller)
+});
+
+/// BoardShim is a primary interface to all boards
+pub struct BoardShim {
+    board_id: BoardId,
+    input_params: BrainFlowInputParams,
+    json_brainflow_input_params: CString,
+}
+
+impl BoardShim {
+    /// Creates a new [BoardShim].
+    pub fn new(board_id: BoardId, input_params: BrainFlowInputParams) -> Result<Self> {
+        let json_input_params = serde_json::to_string(&input_params)?;
+        let json_input_params = CString::new(json_input_params)?;
+        Ok(Self {
+            board_id,
+            input_params,
+            json_brainflow_input_params: json_input_params,
+        })
+    }
+    /// Prepare streaming sesssion and initialize resources.
+    /// You need to call it before any other BoardShim object methods.
+    pub fn prepare_session(&self) -> Result<()> {
+        let res = unsafe {
+            BOARD_CONTROLLER.lock().unwrap().prepare_session(
```

Comment:
One question about lock and mutexes. Inside C\C++ code there is a mutex for such methods and they are thread-safe. So, do we really need it here? If its to ensure that the library is loaded only in one thread then it's ok to have it

Reply:
Due to the dynamic loading of the libraries I need to create a global object, otherwise it wouldn't be possible to use it in the static functions. And since it has to be mutable the mutex is necessary here.

I must say this whole dynamic loading at runtime is only because of the 4 logging functions that share the same name...it would be much easier and nicer if there wouldn't be the name clash i.e. if the log functions would already have a distinctive name - the one they get in the language bindings anyway e.g. enable_logger -> enable_ml_logger